### ISSUE TITLE
Add static analysis

### ISSUE THREAD

**Issue by szepeviktor** [infinum/eightshift-boilerplate#157]
If a developer wants to have robust code here is @phpstan for him/her with WordPress extensions
https://packagist.org/packages/szepeviktor/phpstan-wordpress

- in the /example directory you may find composer.json
- and phpstan config file

**Comment by dingo-d** [comment]
Hi @szepeviktor!

We are already using the [PHP_CodeSniffer](https://github.com/squizlabs/PHP_CodeSniffer/) for code analysis.

I will try it out and see if it gives any major benefits to the project 🙂

**Comment by szepeviktor** [comment]
phpstan analysis the types of all things in the source code - you may know that it is the best static analysis tool for PHP

**Comment by szepeviktor** [comment]
@dingo-d See the result of the boilerplate itself

```
viktor@mail:~/tmp/wp-boilerplate$ vendor/bin/phpstan analyze -l 4
Note: Using configuration file /home/viktor/tmp/wp-boilerplate/phpstan.neon.dist.
 12/12 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

 ------ ------------------------------------------------------------------------------------------------------
  Line   admin/class-admin.php
 ------ ------------------------------------------------------------------------------------------------------
  102    Method Inf_Theme\Admin\Admin::set_admin_color_based_on_env() should return string but returns false.
 ------ ------------------------------------------------------------------------------------------------------

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------
  Line   admin/menu/class-bem-menu-walker.php                                                                                                                
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------
  31     Access to an undefined property Inf_Theme\Admin\Menu\Bem_Menu_Walker::$css_class_prefix.                                                            
  34     Access to an undefined property Inf_Theme\Admin\Menu\Bem_Menu_Walker::$item_css_class_suffixes.                                                     
  59     Parameter #1 $element (array) of method Inf_Theme\Admin\Menu\Bem_Menu_Walker::display_element() should be compatible with parameter $element        
         (object) of method Walker::display_element()                                                                                                        
  59     Parameter #3 $max_depth (array) of method Inf_Theme\Admin\Menu\Bem_Menu_Walker::display_element() should be compatible with parameter $max_depth    
         (int) of method Walker::display_element()                                                                                                           
  59     Parameter #6 $output (array) of method Inf_Theme\Admin\Menu\Bem_Menu_Walker::display_element() should be compatible with parameter $output (string) 
         of method Walker::display_element()                                                                                                                 
  59     Return typehint of method Inf_Theme\Admin\Menu\Bem_Menu_Walker::display_element() has invalid type Inf_Theme\Admin\Menu\element.                    
  81     Parameter #1 $output (array) of method Inf_Theme\Admin\Menu\Bem_Menu_Walker::start_lvl() should be compatible with parameter $output (string) of    
         method Walker_Nav_Menu::start_lvl()                                                                                                                 
  81     Parameter #3 $args (array) of method Inf_Theme\Admin\Menu\Bem_Menu_Walker::start_lvl() should be compatible with parameter $args (stdClass) of      
         method Walker_Nav_Menu::start_lvl()                                                                                                                 
  87     Access to an undefined property Inf_Theme\Admin\Menu\Bem_Menu_Walker::$css_class_prefix.                                                            
  88     Access to an undefined property Inf_Theme\Admin\Menu\Bem_Menu_Walker::$item_css_class_suffixes.                                                     
  98     Binary operation ".=" between array and string results in an error.                                                                                 
  113    Parameter #1 $output (array) of method Inf_Theme\Admin\Menu\Bem_Menu_Walker::start_el() should be compatible with parameter $output (string) of     
         method Walker_Nav_Menu::start_el()                                                                                                                  
  113    Parameter #2 $item (array) of method Inf_Theme\Admin\Menu\Bem_Menu_Walker::start_el() should be compatible with parameter $item (WP_Post) of method 
         Walker_Nav_Menu::start_el()                                                                                                                         
  113    Parameter #4 $args (array) of method Inf_Theme\Admin\Menu\Bem_Menu_Walker::start_el() should be compatible with parameter $args (stdClass) of method
         Walker_Nav_Menu::start_el()                                                                                                                         
  119    Access to an undefined property Inf_Theme\Admin\Menu\Bem_Menu_Walker::$css_class_prefix.                                                            
  120    Access to an undefined property Inf_Theme\Admin\Menu\Bem_Menu_Walker::$item_css_class_suffixes.                                                     
  138    Cannot access property $has_children on array.                                                                                                      
  139    Cannot access property $classes on array.                                                                                                           
  140    Cannot access property $classes on array.                                                                                                           
  141    Cannot access property $classes on array.                                                                                                           
  143    Cannot access property $object_id on array.                                                                                                         
  151    Binary operation ".=" between array and string results in an error.                                                                                 
  178    Cannot access property $before on array.                                                                                                            
  180    Cannot access property $link_before on array.                                                                                                       
  181    Cannot access property $ID on array.                                                                                                                
  181    Cannot access property $title on array.                                                                                                             
  182    Cannot access property $link_after on array.                                                                                                        
  183    Cannot access property $after on array.                                                                                                             
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------

 ------ ------------------------------------------------------------------------------------------------------------------------------------
  Line   admin/menu/class-menu.php
 ------ ------------------------------------------------------------------------------------------------------------------------------------
  67     Parameter $css_class_modifiers of method Inf_Theme\Admin\Menu\Menu::bem_menu() has invalid typehint type Inf_Theme\Admin\Menu\arr.
  72     Call to function is_array() with Inf_Theme\Admin\Menu\arr|string will always evaluate to false.
  84     Variable $modifiers might not be defined.
 ------ ------------------------------------------------------------------------------------------------------------------------------------

 ------ -----------------------------------------------------------------------------------------------------------------------------------------
  Line   class-main.php
 ------ -----------------------------------------------------------------------------------------------------------------------------------------
  48     Method Inf_Theme\Core\Main::get_service_classes() should return array<string> but returns array<int|string, array<int, string>|string>.
 ------ -----------------------------------------------------------------------------------------------------------------------------------------
```

**Comment by dingo-d** [comment]
Looks interesting! I'll definitely try it out. Do you maybe have a PR ready? That would be great so that I can test it as soon as possible and so that we start fixing these issues.

**Comment by iruzevic** [comment]
I have added phpstan to on this [commit](https://github.com/infinum/eightshift-boilerplate/commit/2e69587de754ee63bb2921cd09e413c55fe18115).

We will fix all bugs and push this upgrade on v4.0.1
